servicetalk-concurrent-api-internal/src/test/java/io/servicetalk/concurrent/api/completable/AbstractPublishAndSubscribeOnTest.java

```diff
@@ -15,77 +15,284 @@
+import org.hamcrest.TypeSafeMatcher;
 import org.junit.Rule;
-import org.junit.rules.Timeout;
 
+import java.util.concurrent.Callable;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.RunnableFuture;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReferenceArray;
 import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
-import static io.servicetalk.concurrent.api.Completable.completed;
-import static io.servicetalk.concurrent.api.Completable.never;
+import static io.servicetalk.concurrent.api.Executors.from;
 import static java.lang.Thread.currentThread;
+import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
 
 public abstract class AbstractPublishAndSubscribeOnTest {
-    static final int ORIGINAL_SUBSCRIBER_THREAD = 0;
-    static final int OFFLOADED_SUBSCRIBER_THREAD = 1;
 
+    protected static final String APP_EXECUTOR_PREFIX = "app";
+    protected static final String OFFLOAD_EXECUTOR_PREFIX = "offload";
+
+    static final int APP_THREAD = 0;
+    static final int ON_SUBSCRIBE_THREAD = 1;
+    static final int TERMINAL_THREAD = 2;
+
+    // @Rule
+    // public final Timeout timeout = new ServiceTalkTestTimeout();
     @Rule
-    public final Timeout timeout = new ServiceTalkTestTimeout();
+    public final ExecutorRule<Executor> app = ExecutorRule.withNamePrefix(APP_EXECUTOR_PREFIX);
+    protected Matcher<Thread> offloadExecutor = matchPrefix(OFFLOAD_EXECUTOR_PREFIX);
+    Matcher<Thread> appExecutor = matchPrefix(APP_EXECUTOR_PREFIX);
+    AtomicInteger offloadsStarted = new AtomicInteger();
+    AtomicInteger offloadsFinished = new AtomicInteger();
+    private volatile Runnable afterOffload;
+    private final ThreadPoolExecutor offloadExecutorService = new ThreadPoolExecutor(0, Integer.MAX_VALUE,
+            60L, TimeUnit.SECONDS,
+            new SynchronousQueue<Runnable>(),
+            new DefaultThreadFactory(OFFLOAD_EXECUTOR_PREFIX)) {
+
+        @Override
+        protected void beforeExecute(final Thread t, final Runnable r) {
+            super.beforeExecute(t, r);
+            offloadsStarted.getAndIncrement();
+        }
+
+        @Override
+        protected void afterExecute(final Runnable r, final Throwable t) {
+            offloadsFinished.getAndIncrement();
+            super.afterExecute(r, t);
+        }
+
+        @Override
+        protected <T> RunnableFuture<T> newTaskFor(final Runnable runnable, final T value) {
+            Runnable after = () -> {
+                try {
+                    runnable.run();
+                } finally {
+                    System.out.println("offload finished");
```

Comment:
Consider using logger instead of `System.out`

Reply:
Removed instead